mirrord/agent/src/steal/ip_tables.rs

```diff
@@ -45,258 +75,117 @@ impl IPTables for iptables::IPTables {
         Ok(())
     }
 
+    #[tracing::instrument(level = "trace", skip(self))]
     fn add_rule(&self, chain: &str, rule: &str) -> Result<()> {
         self.append(IPTABLES_TABLE_NAME, chain, rule)
             .map_err(|e| AgentError::IPTablesError(e.to_string()))
     }
 
+    #[tracing::instrument(level = "trace", skip(self))]
     fn insert_rule(&self, chain: &str, rule: &str, index: i32) -> Result<()> {
         self.insert(IPTABLES_TABLE_NAME, chain, rule, index)
             .map_err(|e| AgentError::IPTablesError(e.to_string()))
     }
 
+    #[tracing::instrument(level = "trace", skip(self))]
     fn list_rules(&self, chain: &str) -> Result<Vec<String>> {
         self.list(IPTABLES_TABLE_NAME, chain)
             .map_err(|e| AgentError::IPTablesError(e.to_string()))
     }
 
+    #[tracing::instrument(level = "trace", skip(self))]
     fn remove_rule(&self, chain: &str, rule: &str) -> Result<()> {
         self.delete(IPTABLES_TABLE_NAME, chain, rule)
             .map_err(|e| AgentError::IPTablesError(e.to_string()))
     }
 }
 
+#[enum_dispatch(Redirect)]
+pub enum Redirects<IPT: IPTables + Send + Sync> {
+    Standard(PreroutingRedirect<IPT>),
+    Mesh(MeshRedirect<IPT>),
+    FlushConnections(FlushConnections<Redirects<IPT>>),
+}
+
 /// Wrapper struct for IPTables so it flushes on drop.
-pub(crate) struct SafeIpTables<IPT: IPTables> {
-    inner: IPT,
-    chain_name: String,
-    formatter: IPTableFormatter,
-    flush_connections: bool,
+pub(crate) struct SafeIpTables<IPT: IPTables + Send + Sync> {
+    redirect: Redirects<IPT>,
 }
 
-const IPTABLES_TABLE_NAME: &str = "nat";
 /// Wrapper for using iptables. This creates a a new chain on creation and deletes it on drop.
 /// The way it works is that it adds a chain, then adds a rule to the chain that returns to the
 /// original chain (fallback) and adds a rule in the "PREROUTING" table that jumps to the new chain.
 /// Connections will go then PREROUTING -> OUR_CHAIN -> IF MATCH REDIRECT -> IF NOT MATCH FALLBACK
 /// -> ORIGINAL_CHAIN
 impl<IPT> SafeIpTables<IPT>
 where
-    IPT: IPTables,
+    IPT: IPTables + Send + Sync,
 {
-    pub(super) fn new(ipt: IPT, flush_connections: bool) -> Result<Self> {
-        let formatter = IPTableFormatter::detect(&ipt)?;
-
-        let chain_name = Self::get_chain_name();
-
-        ipt.create_chain(&chain_name)?;
+    pub(super) async fn create(ipt: IPT, flush_connections: bool) -> Result<Self> {
+        let redirect = if let Some(vendor) = MeshVendor::detect(&ipt)? {
+            Redirects::Mesh(MeshRedirect::create(Arc::new(ipt), vendor)?)
+        } else {
+            Redirects::Standard(PreroutingRedirect::create(Arc::new(ipt))?)
+        };
 
-        ipt.add_rule(formatter.entrypoint(), &format!("-j {}", &chain_name))?;
+        let redirect = if flush_connections {
+            Redirects::FlushConnections(FlushConnections::new(Box::new(redirect)))
+        } else {
+            redirect
```

Comment:
A bit weird, can you replace this with the if clause above? i.e. `if let Some(vendor) = MeshVendor::detect(&ipt)?...`

Reply:
I changed it a bit to not have the else clause, also experimented with .with_flush_connections() but I'm not sure about it being needed